hphp/util/asm-x64-intelxed.h

```diff
@@ -0,0 +1,1066 @@
+    return xed_reg(xedFromReg(reg));
+  }
+
+  xed_encoder_operand_t toXedOperand(xed_reg_enum_t reg) {
+    return xed_reg(reg);
+  }
+
+  xed_encoder_operand_t toXedOperand(const MemoryRef& m, int memSize) {
+    xed_reg_enum_t base = int(m.r.base) != -1 ?
+                          xedFromReg(m.r.base) : XED_REG_INVALID;
+    xed_reg_enum_t index = int(m.r.index) != -1 ?
+                           xedFromReg(m.r.index) : XED_REG_INVALID;
+    return xed_mem_gbisd(xed_reg_enum_t(m.segment), base, index, m.r.scale,
+                         xedDispFromValue(m.r.disp), sizeToBits(memSize));
+  }
+
+  xed_encoder_operand_t toXedOperand(const RIPRelativeRef& r, int memSize) {
+    return xed_mem_bd(XED_REG_RIP, xedDispFromValue(r.r.disp, sz::dword),
+                      sizeToBits(memSize));
+  }
+
+  template<typename immtype>
+  xed_encoder_operand_t toXedOperand(const immtype& immed, int immSize) {
+    return xed_imm0(XedImmValue(immed, immSize).uq, sizeToBits(immSize));
+  }
+
+  xed_encoder_operand_t toXedOperand(CodeAddress address, int size) {
+    return xed_relbr(safe_cast<int32_t>((int64_t)address), sizeToBits(size));
+  }
+
+  template<typename immtype>
+  xed_encoder_operand_t toXedOperand(const immtype& immed,
+                                     int immSizes, immFitFunc func) {
+    immSizes = reduceImmSize(immed.q(), immSizes, func);
+    return xed_imm0(XedImmValue(immed, immSizes).uq, sizeToBits(immSizes));
+  }
+
+  int reduceImmSize(int64_t value, int allowedSizes, immFitFunc func) {
+    for (int crtSize = sz::byte; crtSize < sz::qword; crtSize <<= 1) {
+      if ((allowedSizes & crtSize) && (*func)(value, crtSize)) return crtSize;
+    }
+    assertx((allowedSizes & sz::qword) &&
+            "Could not find an optimal size for Immed");
+    return sz::qword;
+  }
+
+  /*
+   * Cache sizes for instruction types in a certain xedInstr context.
+   * This helps with emitting instructions where you need to know in advance
+   * the length of the instruction being emitted (such as when one of
+   * the operands is a RIPRelativeRef) by caching the size of the instruction
+   * and removing the need to call xedEmit twice each time (once to get
+   * the size, and once to actually emit the instruction).
+   */
+  typedef tbb::concurrent_unordered_map<int32_t, uint32_t> XedLenCache;
+
+  ALWAYS_INLINE
+  uint32_t xedCacheLen(XedLenCache* lenCache,
+                       std::function<uint32_t(void)> xedFunc, uint32_t key) {
+    auto res = lenCache->find(key);
+    if (res != lenCache->end()) {
+      return res->second;
+    }
+    auto instrLen = xedFunc();
+    lenCache->insert({key, instrLen});
+    return instrLen;
+  }
+
+  static constexpr uint32_t xedLenCacheKey(xed_iclass_enum_t instr,
+                                           uint32_t size) {
+    // 16 bits should fit a xed_iclass_enum_t value (there are currently ~1560
+    // distinct values).
+    return uint32_t(instr) | (size << 16);
```

Comment:
`size` will completely represent any potential differences in the size of immediates?

Reply:
The cache stores instructions sizes and each entry is uniquely identified by: the instruction itself, and the size of the immediate (which is only used for branch-relative jumps). There's a cache for each xedInstr* function that requires it, which means that the each cached size is defined by: the instruction, the immediate size (in some cases), and the instruction's argument types (since there's a cache for each xedInstr* function which means there's a cache for each combination of instruction argument types) - which should be sufficient.